cedar-policy-validator/src/entity_manifest.rs

```diff
@@ -0,0 +1,1103 @@
+    #[serde(skip_serializing, skip_deserializing)]
+    #[serde(bound(deserialize = "T: Default"))]
+    pub data: T,
+}
+
+/// A data path that may end with requesting the parents of
+/// an entity.
+#[derive(Debug, Clone, PartialEq, Eq)]
+struct AccessPath {
+    /// The root variable that begins the data path
+    pub root: EntityRoot,
+    /// The path of fields of entities or structs
+    pub path: Vec<SmolStr>,
+    /// Request all the parents in the entity hierarchy of this entity.
+    pub parents_required: bool,
+}
+
+/// Entity manifest computation does not handle the full
+/// cedar language. In particular, the policies must follow the
+/// following grammar:
+/// ```text
+/// <expr> = <datapath-expr>
+///          <datapath-expr> in <expr>
+///          <expr> + <expr>
+///          if <expr> { <expr> } { <expr> }
+///          ... all other cedar operators not mentioned by datapath-expr
+
+/// <datapath-expr> = <datapath-expr>.<field>
+///                   <datapath-expr> has <field>
+///                   <variable>
+///                   <entity literal>
+/// ```
+/// The `get_expr_path` function handles `datapath-expr` expressions.
+/// This error message tells the user not to use certain operators
+/// before accessing record or entity attributes, breaking this grammar.
+#[derive(Debug, Clone, Error, Hash, Eq, PartialEq)]
+#[error("For policy `{policy_id}`, failed to analyze expression while computing entity manifest.`")]
+pub struct FailedAnalysisError {
+    /// Source location
+    pub source_loc: Option<Loc>,
+    /// Policy ID where the error occurred
+    pub policy_id: PolicyID,
+    /// The kind of the expression that was unexpected
+    pub expr_kind: ExprKind<Option<Type>>,
+}
+
+impl Diagnostic for FailedAnalysisError {
+    impl_diagnostic_from_source_loc_field!();
+
+    fn help<'a>(&'a self) -> Option<Box<dyn Display + 'a>> {
+        Some(Box::new(format!(
+            "Entity slicing failed to analyze expression: {} operators are not allowed before accessing record or entity attributes.",
+            self.expr_kind.operator_description()
+        )))
+    }
+}
+
+/// An error generated by entity slicing.
+/// See [`FailedAnalysisError`] for details on the fragment
+/// of Cedar handled by entity slicing.
+#[derive(Debug, Error, Diagnostic)]
+#[non_exhaustive]
+pub enum EntityManifestError {
+    /// A validation error was encountered
+    #[error(transparent)]
+    #[diagnostic(transparent)]
+    ValidationError(#[from] ValidationError),
+    /// A entities error was encountered
+    #[error(transparent)]
+    #[diagnostic(transparent)]
+    EntitiesError(#[from] EntitiesError),
+
+    /// The request was partial
+    #[error("Entity slicing requires a fully concrete request. Got a partial request.")]
+    PartialRequestError,
+    /// A policy was partial
+    #[error(
+        "Entity slicing requires fully concrete policies. Got a policy with an unknown expression."
+    )]
+    PartialExpressionError,
```

Comment:
So that we can extend these in the future backwards-compatibly, let's make structs `PartialRequestError` and `PartialExpressionError` that are empty, and do `PartialRequestError(#[from] PartialRequestError)` here

Reply:
done